sentry_streams/sentry_streams/adapters/arroyo/reduce.py

```diff
@@ -0,0 +1,304 @@
+import logging
+import time
+from datetime import timedelta
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    MutableMapping,
+    Optional,
+    Self,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.processing.strategies.reduce import Reduce
+from arroyo.types import BaseValue, FilteredPayload, Message, Partition, Value
+
+from sentry_streams.adapters.arroyo.routes import Route, RoutedValue
+from sentry_streams.pipeline.function_template import Accumulator
+from sentry_streams.pipeline.window import (
+    MeasurementUnit,
+    SlidingWindow,
+    TumblingWindow,
+    Window,
+)
+
+TPayload = TypeVar("TPayload")
+TResult = TypeVar("TResult")
+
+logger = logging.getLogger(__name__)
+
+
+class ArroyoAccumulator:
+
+    def __init__(
+        self,
+        acc: Callable[[], Accumulator[Any, Any]],
+    ) -> None:
+        self.acc = acc
+
+    def initial_value(self) -> Any:
+        # instantaiate the underlying accumulator every time
+        self.instance = self.acc()
+
+        # get the fresh initial value each time
+        return self.instance.get_value()
+
+    # types still tbd
+    def accumulator(self, result: Any, value: BaseValue[RoutedValue]) -> RoutedValue:
+        self.instance.add(value.payload.payload)
+
+        routed = RoutedValue(
+            route=value.payload.route,
+            payload=self.instance.get_value(),
+        )
+
+        return routed
+
+
+class KafkaAccumulator:
+    """
+    Does internal bookkeeping of offsets and timestamps,
+    as well as all the Accumulator functions.
+    """
+
+    def __init__(self, acc: Callable[[], Accumulator[Any, Any]]):
+        self.acc = acc()
+        self.offsets: MutableMapping[Partition, int] = {}
+        self.timestamp = time.time()
+
+    def add(self, value: Any) -> Self:
+
+        payload = value.payload.payload
+        offsets: MutableMapping[Partition, int] = value.committable
+        self.acc.add(payload)
+
+        self.offsets.update(offsets)
+        return self
+
+    def get_value(self) -> Any:
+
+        return self.acc.get_value()
+
+    def merge(self, other: Self) -> Self:
+
+        for partition in other.offsets:
+            if partition in self.offsets:
+                self.offsets[partition] = max(self.offsets[partition], other.offsets[partition])
```

Comment:
Offset committing strategy needs to be changed, I just realized:

When we flush a window, we will commit offsets. The offset to commit will be the highest offset inside of the first Accumulator in the window that is being flushed.